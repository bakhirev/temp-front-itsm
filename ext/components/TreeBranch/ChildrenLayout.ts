import styled from 'styled-components';

import type { Size } from './TreeBranchComponent';

const PADDING_LEFT = {
  big: 40,
  small: 36,
};

interface IChildrenLayoutProps {
  size: Size;
}

export const ChildrenLayout = styled.div<IChildrenLayoutProps>`
  padding-left: ${({ size }) => PADDING_LEFT[size]}px;
`;
