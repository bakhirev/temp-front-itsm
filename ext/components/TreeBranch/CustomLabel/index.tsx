import React from 'react';
import type { FC, ReactElement } from 'react';

import { Subtitle2, Subtitle3, Body1Short, Body2Short } from '../../Typography';
import type { Size } from '../TreeBranchComponent';

import { CustomLabelComponent } from './CustomLabelComponent';
import { IconLayout } from './IconLayout';

interface ICustomLabelProps {
  header?: boolean;
  icon?: ReactElement;
  size: Size;
}

export const CustomLabel: FC<ICustomLabelProps> = ({ children, header, icon, size }) => {
  const getLabel = () => {
    if (header) return size === 'big' ? Subtitle2 : Subtitle3;
    return size === 'big' ? Body1Short : Body2Short;
  };

  const Label = getLabel();

  return (
    <CustomLabelComponent>
      {icon && <IconLayout>{icon}</IconLayout>}
      <Label>{children}</Label>
    </CustomLabelComponent>
  );
};
