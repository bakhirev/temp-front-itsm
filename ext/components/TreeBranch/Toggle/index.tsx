import React from 'react';
import type { FC, KeyboardEvent, MouseEvent } from 'react';

import { KEY_CODES } from '../../common';
import { ReactComponent as ChevronDownOutline } from '../../Icons/system/ChevronDownOutline.svg';
import { ReactComponent as ChevronRightOutline } from '../../Icons/system/ChevronRightOutline.svg';
import type { Size } from '../TreeBranchComponent';

import { ToggleComponent } from './ToggleComponent';

const ICON_SIZE = 24;
const ICON_SIZE_SMALL = 20;

interface IToggleProps {
  expanded?: boolean;
  size: Size;
  onChange: () => void;
}

export const Toggle: FC<IToggleProps> = ({ expanded, size, onChange }) => {
  const handleFocusPrevent = (event: MouseEvent<HTMLDivElement>) => event.preventDefault();
  const handleKeyDown = (event: KeyboardEvent<HTMLDivElement>) => {
    if (event.keyCode === KEY_CODES.ENTER || event.keyCode === KEY_CODES.SPACEBAR) {
      event.preventDefault();
      onChange();
    }
  };

  const Icon = expanded ? ChevronDownOutline : ChevronRightOutline;
  const iconSize = size === 'big' ? ICON_SIZE : ICON_SIZE_SMALL;

  return (
    <ToggleComponent
      size={size}
      tabIndex={1}
      onClick={onChange}
      onKeyDown={handleKeyDown}
      onMouseDown={handleFocusPrevent}
    >
      <Icon height={iconSize} width={iconSize} />
    </ToggleComponent>
  );
};
