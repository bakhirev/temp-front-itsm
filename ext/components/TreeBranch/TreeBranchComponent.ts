import styled from 'styled-components';

export type Size = 'big' | 'small';

const HEIGHT = {
  big: 40,
  small: 32,
};

export const TreeBranchComponent = styled.div<{ size: Size }>`
  align-items: center;
  display: flex;
  height: ${({ size }) => HEIGHT[size]}px;
`;
