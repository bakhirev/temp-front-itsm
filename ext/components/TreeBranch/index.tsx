import React from 'react';
import type { FC, ReactElement } from 'react';

import { Checkbox } from '../Checkbox';

import { ChildrenLayout } from './ChildrenLayout';
import { CustomLabel } from './CustomLabel';
import { Toggle } from './Toggle';
import { TreeBranchComponent } from './TreeBranchComponent';
import type { Size } from './TreeBranchComponent';

export interface ITreeBranchProps {
  /** Состояние чекбокса: выбран/не выбран */
  checked?: boolean;
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
  /** Показывать/скрывать дочернии ветки */
  expanded?: boolean;
  /** Отображение ветки как заголовка (жирный текст) */
  header?: boolean;
  /** Скрыть чекбокс и отображать только текст (и иконку) */
  hideCheckbox?: boolean;
  /** Иконка, которая будет отображаться слева от текста */
  icon?: ReactElement;
  /** Уникальный идентификатор ветки */
  id: string | number;
  /** Неопределённое состояние чекбокса */
  indeterminate?: boolean;
  /** Текст, который будет отображаться в ветке */
  label: string;
  /** Размер ветки */
  size?: Size;
  /** Коллбэк на изменение состояния чекбокса. Срабатывает при клике или нажатии Enter/Space при переходе по табу */
  onCheck?: (id: string | number, checked: boolean) => void;
  /** Коллбэк на изменение состояния переключателя, отвечающего за показ/скрытие дочерних веток. Срабатывает при клике или нажатии Enter/Space при переходе по табу */
  onExpand?: (id: string | number, expanded: boolean) => void;
}

export const TreeBranch: FC<ITreeBranchProps> = ({
  checked,
  children,
  className,
  dataTestId,
  expanded,
  header,
  hideCheckbox,
  icon,
  id,
  indeterminate,
  label,
  size = 'big',
  onCheck,
  onExpand,
}) => {
  const renderLabel = () => {
    if (!hideCheckbox && !header && !icon) return label;
    return (
      <CustomLabel header={header} icon={icon} size={size}>
        {label}
      </CustomLabel>
    );
  };

  return (
    <>
      <TreeBranchComponent className={className} data-test-id={dataTestId} size={size}>
        {children && (
          <Toggle expanded={expanded} size={size} onChange={() => onExpand?.(id, !expanded)} />
        )}
        {hideCheckbox ? (
          renderLabel()
        ) : (
          <Checkbox
            checked={checked}
            indeterminate={indeterminate}
            label={renderLabel()}
            size={size}
            onChange={(_, checked) => onCheck?.(id, checked)}
          />
        )}
      </TreeBranchComponent>
      {expanded && children && <ChildrenLayout size={size}>{children}</ChildrenLayout>}
    </>
  );
};
