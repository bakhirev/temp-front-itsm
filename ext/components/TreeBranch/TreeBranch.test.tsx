/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable react/display-name */
import React from 'react';
import { shallow } from 'enzyme';

import { TreeBranch } from '../TreeBranch';
import type { ITreeBranchProps } from '../TreeBranch';

jest.mock('../Checkbox', () => ({
  Checkbox: ({ checked, onChange }) => <div onClick={(event) => onChange(event, !checked)} />,
}));

describe('TreeBranch', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const requiredProps: ITreeBranchProps = {
    id: 0,
    label: 'Foo',
  };

  const childProps: ITreeBranchProps = {
    id: 1,
    label: 'Bar',
  };

  it('should render component with required props', () => {
    const wrapper = shallow(<TreeBranch {...requiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with different sizes', () => {
    const sizes: Array<ITreeBranchProps['size']> = ['big', 'small'];

    sizes.forEach((size) => {
      const wrapper = shallow(<TreeBranch size={size} {...requiredProps} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('should render component as header', () => {
    const wrapper = shallow(<TreeBranch header {...requiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with toggle if children are provided', () => {
    const wrapper = shallow(
      <TreeBranch {...requiredProps}>
        <TreeBranch {...childProps} />
      </TreeBranch>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should show children if expanded prop is true', () => {
    const wrapper = shallow(
      <TreeBranch expanded {...requiredProps}>
        <TreeBranch {...childProps} />
      </TreeBranch>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with icon', () => {
    const Icon = () => <div>Icon</div>;
    const wrapper = shallow(<TreeBranch icon={<Icon />} {...requiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with indeterminate checkbox', () => {
    const wrapper = shallow(<TreeBranch indeterminate {...requiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component without checkbox', () => {
    const wrapper = shallow(<TreeBranch hideCheckbox {...requiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call onCheck with id and new checked when user clicks on checkbox', () => {
    const checked = false;
    const onCheck = jest.fn();
    const wrapper = shallow(<TreeBranch checked={checked} onCheck={onCheck} {...requiredProps} />);

    const checkbox = wrapper.childAt(0).childAt(0).dive();
    checkbox.simulate('click');

    expect(onCheck).toHaveBeenCalledWith(requiredProps.id, !checked);
  });

  it('should call onExpand with id and new expanded when user clicks on toggle', () => {
    const expanded = false;
    const onExpand = jest.fn();
    const wrapper = shallow(
      <TreeBranch expanded={expanded} onExpand={onExpand} {...requiredProps}>
        <TreeBranch {...childProps} />
      </TreeBranch>
    );

    const toggle = wrapper.childAt(0).childAt(0).dive();
    toggle.simulate('click');

    expect(onExpand).toHaveBeenCalledWith(requiredProps.id, !expanded);
  });
});
