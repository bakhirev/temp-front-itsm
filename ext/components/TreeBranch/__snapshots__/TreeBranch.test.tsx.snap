// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeBranch should render component as header 1`] = `
<Fragment>
  <styled.div
    size="big"
  >
    <Checkbox
      label={
        <Unknown
          header={true}
          size="big"
        >
          Foo
        </Unknown>
      }
      onChange={[Function]}
      size="big"
    />
  </styled.div>
</Fragment>
`;

exports[`TreeBranch should render component with different sizes 1`] = `
<Fragment>
  <styled.div
    size="big"
  >
    <Checkbox
      label="Foo"
      onChange={[Function]}
      size="big"
    />
  </styled.div>
</Fragment>
`;

exports[`TreeBranch should render component with different sizes 2`] = `
<Fragment>
  <styled.div
    size="small"
  >
    <Checkbox
      label="Foo"
      onChange={[Function]}
      size="small"
    />
  </styled.div>
</Fragment>
`;

exports[`TreeBranch should render component with icon 1`] = `
<Fragment>
  <styled.div
    size="big"
  >
    <Checkbox
      label={
        <Unknown
          icon={<Icon />}
          size="big"
        >
          Foo
        </Unknown>
      }
      onChange={[Function]}
      size="big"
    />
  </styled.div>
</Fragment>
`;

exports[`TreeBranch should render component with indeterminate checkbox 1`] = `
<Fragment>
  <styled.div
    size="big"
  >
    <Checkbox
      indeterminate={true}
      label="Foo"
      onChange={[Function]}
      size="big"
    />
  </styled.div>
</Fragment>
`;

exports[`TreeBranch should render component with required props 1`] = `
<Fragment>
  <styled.div
    size="big"
  >
    <Checkbox
      label="Foo"
      onChange={[Function]}
      size="big"
    />
  </styled.div>
</Fragment>
`;

exports[`TreeBranch should render component with toggle if children are provided 1`] = `
<Fragment>
  <styled.div
    size="big"
  >
    <Component
      onChange={[Function]}
      size="big"
    />
    <Checkbox
      label="Foo"
      onChange={[Function]}
      size="big"
    />
  </styled.div>
</Fragment>
`;

exports[`TreeBranch should render component without checkbox 1`] = `
<Fragment>
  <styled.div
    size="big"
  >
    <Component
      size="big"
    >
      Foo
    </Component>
  </styled.div>
</Fragment>
`;

exports[`TreeBranch should show children if expanded prop is true 1`] = `
<Fragment>
  <styled.div
    size="big"
  >
    <Component
      expanded={true}
      onChange={[Function]}
      size="big"
    />
    <Checkbox
      label="Foo"
      onChange={[Function]}
      size="big"
    />
  </styled.div>
  <styled.div
    size="big"
  >
    <Component
      id={1}
      label="Bar"
    />
  </styled.div>
</Fragment>
`;
